rig-core/src/embeddings/embeddable.rs

```diff
@@ -0,0 +1,359 @@
+//! The module defines the [Embeddable] trait, which must be implemented for types that can be embedded.
+//! //! # Example
+//! ```rust
+//! use std::env;
+//!
+//! use serde::{Deserialize, Serialize};
+//!
+//! struct FakeDefinition {
+//!     id: String,
+//!     word: String,
+//!     definitions: Vec<String>,
+//! }
+//!
+//! let fake_definition = FakeDefinition {
+//!     id: "doc1".to_string(),
+//!     word: "hello".to_string(),
+//!     definition: "used as a greeting or to begin a conversation".to_string()
+//! };
+//!
+//! impl Embeddable for FakeDefinition {
+//!     type Error = anyhow::Error;
+//!
+//!     fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error> {
+//!         // Embeddigns only need to be generated for `definition` field.
+//!         // Select it from the struct and return it as a single item.
+//!         Ok(OneOrMany::one(self.definition.clone()))
+//!     }
+//! }
+//! ```
+
+/// Error type used for when the `embeddable` method fails.
+/// Used by default implementations of `Embeddable` for common types.
+#[derive(Debug, thiserror::Error)]
+pub enum EmbeddableError {
+    #[error("SerdeError: {0}")]
+    SerdeError(#[from] serde_json::Error),
+}
+
+/// Trait for types that can be embedded.
+/// The `embeddable` method returns a OneOrMany<String> which contains strings for which embeddings will be generated by the embeddings builder.
+/// If there is an error generating the list of strings, the method should return an error that implements `std::error::Error`.
+pub trait Embeddable {
+    type Error: std::error::Error;
+
+    fn embeddable(&self) -> Result<OneOrMany<String>, Self::Error>;
+}
+
+/// Struct containing either a single item or a list of items of type T.
+/// If a single item is present, `first` will contain it and `rest` will be empty.
+/// If multiple items are present, `first` will contain the first item and `rest` will contain the rest.
+/// IMPORTANT: this struct cannot be created with an empty vector.
+/// OneOrMany objects can only be created using OneOrMany::one() or OneOrMany::many().
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct OneOrMany<T> {
+    /// First item in the list.
+    first: T,
+    /// Rest of the items in the list.
+    rest: Vec<T>,
+}
+
+impl<T: Clone> OneOrMany<T> {
+    /// Get the first item in the list.
+    pub fn first(&self) -> T {
+        self.first.clone()
+    }
+
+    /// Get the rest of the items in the list (excluding the first one).
+    pub fn rest(&self) -> Vec<T> {
+        self.rest.clone()
+    }
+
```

Comment:
Note: I created the methods `one` and `many` to create objects of type `OneOrMany` instead of doing 
`impl From<T> for OneOrMany<T>` and `impl From<Vec<T>> for OneOrMany<T>` because these are conflicting. See my comment below for why I would use  `impl From<Vec<T>> for OneOrMany<T>` instead of  `impl TryFrom<Vec<T>> for OneOrMany<T>`.